In [1]:
import graphlab

In [2]:
song_data = graphlab.SFrame('song_data.gl/')
graphlab.canvas.set_target('ipynb')
song_data.print_rows(num_rows=3)
print 'len of song_data = ', len(song_data)
print 'len of unique user id= ', len(song_data['user_id'].unique())
print 'len of unique song = ', len(song_data['song'].unique())

This non-commercial license of GraphLab Create for academic use is assigned to r38411@yahoo.com and will expire on August 31, 2019.


[INFO] graphlab.cython.cy_server: GraphLab Create v2.1 started. Logging: /tmp/graphlab_server_1538576399.log


+-------------------------------+--------------------+--------------+
|            user_id            |      song_id       | listen_count |
+-------------------------------+--------------------+--------------+
| b80344d063b5ccb3212f76538f... | SOAKIMP12A8C130995 |      1       |
| b80344d063b5ccb3212f76538f... | SOBBMDR12A8C13253B |      2       |
| b80344d063b5ccb3212f76538f... | SOBXHDL12A81C204C0 |      1       |
+-------------------------------+--------------------+--------------+
+-----------------+---------------+-------------------------------+
|      title      |     artist    |              song             |
+-----------------+---------------+-------------------------------+
|     The Cove    |  Jack Johnson |    The Cove - Jack Johnson    |
| Entre Dos Aguas | Paco De Lucia | Entre Dos Aguas - Paco De ... |
|     Stronger    |   Kanye West  |     Stronger - Kanye West     |
+-----------------+---------------+-------------------------------+
[1116609 rows x 6 columns]

len of

In [3]:
## question 1: find out the # of unique users for each of the artist: Kanye West, Foo FIghters, Taylor Swift, Lady GaGa

In [4]:
artist = ['Kanye West', 'Foo Fighters', 'Taylor Swift', 'Lady GaGa']

In [5]:
for i in artist:
    user_artist = song_data[song_data['artist'] == i]
    print 'for ', i, ' , # of unique users are = ', len(user_artist['user_id'].unique())

for  Kanye West  , # of unique users are =  2522
for  Foo Fighters  , # of unique users are =  2055
for  Taylor Swift  , # of unique users are =  3246
for  Lady GaGa  , # of unique users are =  2928


# Question 1:
# Kayne West, num of unique users are: 2522
# Foo Fighters, 2055
# Taylor Swift: 3246
# Lady Gaga: 2928

In [6]:
## Question 2: find out the most popular and least popular artist

In [7]:
artist_count = song_data.groupby(key_columns='artist', 
                                 operations={'total_count': graphlab.aggregate.SUM('listen_count')})
artist_count = artist_count.sort('total_count', ascending=False)
print artist_count[0:5]
artist_count.tail()

+------------------------+-------------+
|         artist         | total_count |
+------------------------+-------------+
|     Kings Of Leon      |    43218    |
|     Dwight Yoakam      |    40619    |
|         Björk          |    38889    |
|        Coldplay        |    35362    |
| Florence + The Machine |    33387    |
+------------------------+-------------+
[5 rows x 2 columns]



Columns:
	artist	str
	total_count	int

Rows: 10

Data:
+-------------------------------+-------------+
|             artist            | total_count |
+-------------------------------+-------------+
|          Jody Bernal          |      38     |
|        Aneta Langerova        |      38     |
| Kanye West / Talib Kweli /... |      38     |
|             Nâdiya            |      36     |
|         harvey summers        |      31     |
|         Boggle Karaoke        |      30     |
|             Diplo             |      30     |
| Beyoncé feat. Bun B and Sl... |      26     |
|         Reel Feelings         |      24     |
|        William Tabbert        |      14     |
+-------------------------------+-------------+
[10 rows x 2 columns]

# Question 2: 
# The most popular artist = Kings ofLeon: 43218
# The least artist: William Tabbert: 14

In [8]:
## Question 3: build the recommender

In [14]:
train_data, test_data = song_data.random_split(.8, seed=0)

In [16]:
subset_test_users = test_data['user_id'].unique()[0:10000]

In [18]:
personalized_model = graphlab.item_similarity_recommender.create(train_data,
                                                                user_id='user_id',
                                                                item_id='song')

Recsys training: model = item_similarity

Warning: Ignoring columns song_id, listen_count, title, artist;

To use one of these as a target column, set target = <column_name>

and use a method that allows the use of a target.

Preparing data set.

Data has 893580 observations with 66085 users and 9952 items.

Data prepared in: 0.90752s

Training model from provided data.

Gathering per-item and per-user statistics.

+--------------------------------+------------+

| Elapsed Time (Item Statistics) | % Complete |

+--------------------------------+------------+

| 2.936ms                        | 4.5        |

| 43.101ms                       | 100        |

+--------------------------------+------------+

Setting up lookup tables.

Processing data in one pass using dense lookup tables.

+-------------------------------------+------------------+-----------------+

| Elapsed Time (Constructing Lookups) | Total % Complete | Items Processed |

+-------------------------------------+------------------+-----------------+

| 235.591ms                           | 0                | 0               |

| 1.14s                               | 100              | 9952            |

+-------------------------------------+------------------+-----------------+

Finalizing lookup tables.

Generating candidate set for working with new users.

Finished training in 2.20894s

In [22]:
subset_recom = personalized_model.recommend(subset_test_users, k=1)
subset_recom.print_rows(num_rows=5)
print len(subset_recom)

recommendations finished on 1000/10000 queries. users per second: 12682.6

recommendations finished on 2000/10000 queries. users per second: 16435.7

recommendations finished on 3000/10000 queries. users per second: 17195.6

recommendations finished on 4000/10000 queries. users per second: 17959.3

recommendations finished on 5000/10000 queries. users per second: 18322.9

recommendations finished on 6000/10000 queries. users per second: 18827

recommendations finished on 7000/10000 queries. users per second: 19277.3

recommendations finished on 8000/10000 queries. users per second: 19561.1

recommendations finished on 9000/10000 queries. users per second: 19841.6

recommendations finished on 10000/10000 queries. users per second: 20030.1

+-------------------------------+-------------------------------+
|            user_id            |              song             |
+-------------------------------+-------------------------------+
| c66c10a9567f0d82ff31441a9f... | Cuando Pase El Temblor - S... |
| c067c22072a17d33310d7223d7... | Grind With Me (Explicit Ve... |
| f6c596a519698c97f1591ad89f... |    Hey_ Soul Sister - Train   |
| 696787172dd3f5169dc94deef9... | Senza Una Donna (Without A... |
| 3a7111f4cdf3c5a85fd4053e3c... | Heartbreak Warfare - John ... |
+-------------------------------+-------------------------------+
+-----------------+------+
|      score      | rank |
+-----------------+------+
| 0.0194504536115 |  1   |
| 0.0459424376488 |  1   |
| 0.0238929539919 |  1   |
|  0.017026577677 |  1   |
| 0.0298416515191 |  1   |
+-----------------+------+
[10000 rows x 4 columns]

10000


In [24]:
subset_count = subset_recom.groupby(key_columns='song',operations={'total_count':graphlab.aggregate.COUNT()})
subset_count = subset_count.sort('total_count', ascending=False)
subset_count

Columns:
	song	str
	total_count	int

Rows: 3141

Data:
+--------------------------------+-------------+
|              song              | total_count |
+--------------------------------+-------------+
|          Undo - Björk          |     439     |
|     Secrets - OneRepublic      |     355     |
|    Revelry - Kings Of Leon     |     228     |
| You're The One - Dwight Yoakam |     163     |
| Fireflies - Charttraxx Karaoke |     122     |
|    Hey_ Soul Sister - Train    |      99     |
|    Sehr kosmisch - Harmonia    |      99     |
| Horn Concerto No. 4 in E f...  |      96     |
| OMG - Usher featuring will...  |      59     |
| Dog Days Are Over (Radio E...  |      49     |
+--------------------------------+-------------+
[3141 rows x 2 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

# Question 3: The most recommended song to the first 10000 users are:
# Undo-Bjork: total-count = 439

In [30]:
%matplotlib inline
model_performance = graphlab.compare(test_data, [personalized_model], user_sample=0.05)
model_performance

compare_models: using 2931 users to estimate model performance
PROGRESS: Evaluate model M0


recommendations finished on 1000/2931 queries. users per second: 14667.7

recommendations finished on 2000/2931 queries. users per second: 16932.5


Precision and recall summary statistics by cutoff
+--------+-----------------+-----------------+
| cutoff |  mean_precision |   mean_recall   |
+--------+-----------------+-----------------+
|   1    |  0.185261003071 | 0.0580295022465 |
|   2    |  0.15626066189  | 0.0961270447964 |
|   3    |  0.136017286478 |  0.120667573073 |
|   4    |  0.12273967929  |  0.142352233422 |
|   5    |  0.113135448652 |  0.161672664308 |
|   6    |  0.102752189241 |  0.174263294043 |
|   7    | 0.0952868353073 |  0.187933118665 |
|   8    | 0.0890054588878 |   0.1998094233  |
|   9    | 0.0841578528375 |  0.212721372982 |
|   10   | 0.0795291709314 |  0.222232492857 |
+--------+-----------------+-----------------+
[10 rows x 3 columns]

Model compare metric: precision_recall


Columns:
	dataset	str
	metric	str
	model	str
	model_type	str
	results	dict

Rows: 1

Data:
+-----------+------------------+--------------------+---------------------------+
|  dataset  |      metric      |       model        |         model_type        |
+-----------+------------------+--------------------+---------------------------+
| test_data | precision_recall | personalized_model | ItemSimilarityRecommender |
+-----------+------------------+--------------------+---------------------------+
+-------------------------------+
|            results            |
+-------------------------------+
| {'precision_recall': [{'cu... |
+-------------------------------+
[1 rows x 5 columns]